File: go-terrapod/deleted_workspaces.go
package terrapod

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
)

// DeletedWorkspace is a workspace that has been deleted but whose state is
// still recoverable (#1253).
//
// Deleting a workspace removes its rows — state-version rows go by CASCADE —
// but not its state blobs. A marker written at delete time is what makes those
// blobs findable again, and what gives the retention reaper something to age;
// this type is that marker as the API serves it.
//
// The window is finite. Once DeletedAt is older than the server's
// deleted_workspace_retention_days the state is reaped and the workspace is
// gone for good, so RestorableUntil is the field a caller should act on.
type DeletedWorkspace struct {
	WorkspaceID   string `json:"workspace-id"`
	WorkspaceName string `json:"workspace-name"`
	DeletedAt     string `json:"deleted-at"`
	DeletedBy     string `json:"deleted-by"`

	// MarkerReason distinguishes how the marker came to exist: "deleted" means
	// it was written by the delete itself and DeletedAt is the true deletion
	// time; "discovered-orphaned" means the reaper found state with no marker
	// (a workspace deleted before this feature shipped, or one whose marker
	// write failed) and DeletedAt is merely when it was first seen.
	MarkerReason string `json:"marker-reason"`

	LastSerial int    `json:"last-serial"`
	Lineage    string `json:"lineage"`

	// StateVersionsAvailable is counted from storage at request time, not
	// taken from the marker — a partial reap or an incomplete replication
	// shows up here and nowhere else.
	StateVersionsAvailable int      `json:"state-versions-available"`
	AgeDays                *float64 `json:"age-days"`

	// RestorableUntil is empty when retention is disabled (0 days), meaning
	// the state is never reaped automatically.
	RestorableUntil string `json:"restorable-until"`

	Settings      map[string]any             `json:"settings"`
	VariableNames []DeletedWorkspaceVariable `json:"variable-names"`

	// RestoredTo lists the workspaces this deletion has already been restored
	// into, empty until the first restore. It is not bookkeeping: a second
	// restore yields a second live workspace holding the SAME state lineage
	// and serial over the same infrastructure, after which an apply in either
	// makes the other's next plan read as wholesale drift. Check this before
	// offering a restore; the server refuses a repeat unless forced.
	RestoredTo []string `json:"restored-to"`
	RestoredAt string   `json:"restored-at"`
	RestoredBy string   `json:"restored-by"`
}

// DeletedWorkspaceVariable names a variable the deleted workspace had.
//
// Names only, never values: the marker is a plain object in the bucket and is
// replicated to any standby, so a value there would be a secret at rest
// outside the encryption boundary. The names are what an operator needs in
// order to know what to recreate after a restore.
type DeletedWorkspaceVariable struct {
	Key       string `json:"key"`
	Category  string `json:"category"`
	Sensitive bool   `json:"sensitive"`
}

// RestoredWorkspace is the outcome of a restore.
//
// The restore produces a NEW workspace rather than reviving the original in
// place, so ID is not RestoredFrom. Suppressed and DroppedReferences describe
// what was deliberately not carried across, and are the point of the response:
// a restored workspace comes back inert and it is the caller's job to decide
// what to switch back on.
type RestoredWorkspace struct {
	ID                    string           `json:"-"` // from the resource id, not attributes
	Name                  string           `json:"name"`
	RestoredFrom          string           `json:"restored-from"`
	StateVersionsRestored int              `json:"state-versions-restored"`
	StateVersionsSkipped  []map[string]any `json:"state-versions-skipped"`

	// Suppressed names settings that were true in the snapshot but forced off:
	// auto-apply, drift detection, and the VCS connection. A workspace that
	// applied the moment it was restored — against infrastructure that may
	// have drifted or been partly destroyed since the delete — is the failure
	// mode this prevents.
	Suppressed []string `json:"suppressed"`

	// DroppedReferences names things that pointed at other resources and were
	// not re-attached, because the id may since have been reused by something
	// entirely different.
	DroppedReferences []map[string]any `json:"dropped-references"`
}

// DeletedWorkspaceListOptions controls paging of the undelete list.
type DeletedWorkspaceListOptions struct {
	PageNumber int
	PageSize   int
}

// DeletedWorkspaceList is one page of deleted workspaces plus its pagination.
type DeletedWorkspaceList struct {
	Items []DeletedWorkspace
	Meta  ListMeta
}

// ListDeletedWorkspaces returns one page of recoverable deleted workspaces,
// newest deletion first. Requires platform admin.
func (c *Client) ListDeletedWorkspaces(ctx context.Context, opts DeletedWorkspaceListOptions) (*DeletedWorkspaceList, error) {
	q := url.Values{}
	if opts.PageNumber > 0 {
		q.Set("page[number]", strconv.Itoa(opts.PageNumber))
	}
	if opts.PageSize > 0 {
		q.Set("page[size]", strconv.Itoa(opts.PageSize))
	}
	path := "/api/terrapod/v1/deleted-workspaces"
	if encoded := q.Encode(); encoded != "" {
		path += "?" + encoded
	}
	body, err := c.Get(ctx, path)
	if err != nil {
		return nil, err
	}
	var doc struct {
		Data []struct {
			Attributes DeletedWorkspace `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, fmt.Errorf("decode deleted workspaces: %w", err)
	}
	out := &DeletedWorkspaceList{Items: make([]DeletedWorkspace, 0, len(doc.Data))}
	for _, d := range doc.Data {
		out.Items = append(out.Items, d.Attributes)
	}
	meta, err := parseListMeta(body)
	if err != nil {
		// Not swallowed (#1301). With meta discarded, TotalPages stays 0 and
		// ListAllDeletedWorkspaces below stops after page one — returning a
		// truncated list as a SUCCESS. For this feature that fails in the
		// worst direction: "there is nothing else to restore" is exactly the
		// answer an operator must not be given wrongly.
		return nil, fmt.Errorf("decode deleted workspaces pagination: %w", err)
	}
	out.Meta = meta
	return out, nil
}

// ListAllDeletedWorkspaces pages through every recoverable deleted workspace.
func (c *Client) ListAllDeletedWorkspaces(ctx context.Context) ([]DeletedWorkspace, error) {
	var all []DeletedWorkspace
	for page := 1; ; page++ {
		list, err := c.ListDeletedWorkspaces(ctx, DeletedWorkspaceListOptions{PageNumber: page, PageSize: 100})
		if err != nil {
			return nil, err
		}
		all = append(all, list.Items...)
		if len(list.Items) == 0 || page >= list.Meta.TotalPages {
			return all, nil
		}
	}
}

// GetDeletedWorkspace fetches one deleted workspace's marker.
func (c *Client) GetDeletedWorkspace(ctx context.Context, workspaceID string) (*DeletedWorkspace, error) {
	body, err := c.Get(ctx, "/api/terrapod/v1/deleted-workspaces/"+workspaceID)
	if err != nil {
		return nil, err
	}
	var doc struct {
		Data struct {
			Attributes DeletedWorkspace `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, fmt.Errorf("decode deleted workspace: %w", err)
	}
	return &doc.Data.Attributes, nil
}

// RestoreOptions tunes a restore.
type RestoreOptions struct {
	// Name for the recovered workspace. Empty reuses the original, suffixed if
	// it has been taken since.
	Name string

	// Force permits restoring a deletion that has already been restored once.
	// Without it the server refuses, because the repeat would produce a second
	// live workspace over the same infrastructure with the same lineage. Set
	// it only when the earlier restore is known to be gone.
	Force bool
}

// RestoreDeletedWorkspace recovers a deleted workspace's state into a new
// workspace. Requires platform admin.
//
// Returns a ConflictError in two cases, both of which are failures rather than
// empty successes so a caller is never handed a bare workspace and told the
// restore worked: no state could be recovered (the retention window has passed
// and the blobs are gone), or this deletion has already been restored — see
// RestoreOptions.Force.
//
// Only the newest state versions are copied, bounded by the server's
// state_versions_keep; anything beyond that is reported in
// RestoredWorkspace.StateVersionsSkipped rather than silently dropped.
func (c *Client) RestoreDeletedWorkspace(ctx context.Context, workspaceID string, opts RestoreOptions) (*RestoredWorkspace, error) {
	attrs := map[string]any{}
	if opts.Name != "" {
		attrs["name"] = opts.Name
	}
	if opts.Force {
		attrs["force"] = true
	}
	payload, err := json.Marshal(map[string]any{
		"data": map[string]any{"type": "workspaces", "attributes": attrs},
	})
	if err != nil {
		return nil, fmt.Errorf("encode restore request: %w", err)
	}
	body, err := c.Post(ctx, "/api/terrapod/v1/deleted-workspaces/"+workspaceID+"/restore", payload)
	if err != nil {
		return nil, err
	}
	var doc struct {
		Data struct {
			ID         string            `json:"id"`
			Attributes RestoredWorkspace `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, fmt.Errorf("decode restored workspace: %w", err)
	}
	out := doc.Data.Attributes
	out.ID = doc.Data.ID
	return &out, nil
}

File: services/terrapod/api/routers/deleted_workspaces.py
"""Undelete surface for deleted workspaces (#1253).

UX CONTRACT: consumed by `web/src/app/admin/deleted-workspaces/page.tsx`.
Changes to response shapes, attribute names, or status codes here MUST be
matched there.

**Platform admin only, on every route including the reads.** A delete marker
names a workspace, its labels and its variable *names*, and restoring one
materialises its state history — which is to say its secrets — into a workspace
the caller can then read. That is a privilege escalation for anyone who did not
have access to the original, so it is not delegated through workspace RBAC:
the original's ACL died with its rows and cannot be consulted.
"""

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession

from terrapod.api.dependencies import AuthenticatedUser, require_admin
from terrapod.api.pagination import paginate
from terrapod.config import settings
from terrapod.db.session import get_db
from terrapod.logging_config import get_logger
from terrapod.services import deleted_workspace_service as dws
from terrapod.services.workspace_name import validate_workspace_name
from terrapod.storage import get_storage

logger = get_logger(__name__)

router = APIRouter(tags=["deleted-workspaces"])


def _marker_json(m: dict) -> dict:
    """Serialise a marker as a JSON:API resource.

    `variable-names` carries names and categories only — never values. That is
    an invariant of the marker itself (see `deleted_workspace_service`), so
    this is a passthrough rather than a filter; it is called out because a
    future field added to the marker must be checked before it lands here.
    """
    return {
        "id": m.get("workspace_id"),
        "type": "deleted-workspaces",
        "attributes": {
            "workspace-id": m.get("workspace_id"),
            "workspace-name": m.get("workspace_name"),
            "deleted-at": m.get("deleted_at"),
            "deleted-by": m.get("deleted_by"),
            "marker-reason": m.get("marker_reason"),
            "last-serial": m.get("last_serial"),
            "lineage": m.get("lineage"),
            "state-versions-available": m.get("state_versions_available"),
            "age-days": m.get("age_days"),
            "restorable-until": m.get("restorable_until"),
            "settings": m.get("settings") or {},
            "variable-names": m.get("variable_names") or [],
            # Empty until this deletion has been restored at least once. A
            # second restore over the same lineage is the failure mode this
            # exists to make visible (#1299).
            "restored-to": [f"ws-{i}" for i in m.get("restored_to") or []],
            "restored-at": m.get("restored_at"),
            "restored-by": m.get("restored_by"),
        },
    }


def _raw_id(workspace_id: str) -> str:
    """Accept a workspace id in either form.

    Marker ids are bare UUIDs — the marker key is `state/deleted/{uuid}.json`,
    written before any serializer sees it — while every other workspace id in
    this API is `ws-`-prefixed. Rejecting the prefixed form made callers keep
    two spellings of the same id straight, and was conspicuous enough that the
    MCP tool description had to warn about it (#1299). Accepting both costs
    nothing; the emitted id stays bare so existing consumers are unaffected.
    """
    return workspace_id.removeprefix("ws-")


@router.get("/deleted-workspaces")
async def list_deleted_workspaces(
    request: Request,
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Deleted workspaces still recoverable, newest deletion first."""
    storage = get_storage()
    markers = await dws.list_deleted(
        db, storage, settings.artifact_retention.deleted_workspace_retention_days
    )
    # Page FIRST, then count state versions: the count costs a full prefix
    # listing per marker, and paying it for every deleted workspace ever before
    # discarding all but one page is what made this endpoint scale with the
    # deployment's whole deletion history (#1299).
    page, meta = paginate(markers, request)
    await dws.attach_state_counts(storage, page)
    return {"data": [_marker_json(m) for m in page], "meta": meta}


@router.get("/deleted-workspaces/{workspace_id}")
async def get_deleted_workspace(
    workspace_id: str,
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """One deleted workspace's marker, with its recoverable state count."""
    marker = await dws.get_deleted(
        db,
        get_storage(),
        _raw_id(workspace_id),
        settings.artifact_retention.deleted_workspace_retention_days,
    )
    if marker is None:
        raise HTTPException(status_code=404, detail="Deleted workspace not found")
    return {"data": _marker_json(marker)}


@router.post("/deleted-workspaces/{workspace_id}/restore", status_code=201)
async def restore_deleted_workspace(
    workspace_id: str,
    payload: dict | None = None,
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Recover a deleted workspace as a new workspace holding its state.

    This deliberately produces a **new** workspace with a new id rather than
    reviving the original in place. Re-attaching the original id would be
    cheaper — the state prefix is already there and no copy would be needed —
    but a frictionless one-click undo makes deletion feel free, and deletion
    should not feel free. Recovery is an explicit operation with a visible
    cost, and the result is honestly labelled as a new workspace.

    Returns the new workspace id plus a report of what was deliberately not
    carried over, so the operator knows what to re-enable.
    """
    workspace_id = _raw_id(workspace_id)
    storage = get_storage()

    name = None
    force = False
    if isinstance(payload, dict):
        attrs = (payload.get("data") or {}).get("attributes") or {}
        force = bool(attrs.get("force"))
        raw = attrs.get("name")
        if raw is not None:
            if not isinstance(raw, str):
                raise HTTPException(status_code=422, detail="name must be a non-empty string")
            # The full format contract, not just "non-empty" (#1299). Every
            # other workspace-creating path enforces it, and the name is
            # load-bearing downstream: it is what the `cloud {}` block matches
            # on, what `/app/{org}/{name}` redirects by, and what appears in
            # the DR state index and VCS status contexts. Strict here because
            # this one came from a human who can simply retype it; the
            # marker-derived fallback sanitizes instead, so a corrupt marker
            # cannot make a recoverable workspace un-restorable.
            try:
                name = validate_workspace_name(raw)
            except ValueError as e:
                raise HTTPException(status_code=422, detail=str(e)) from e

    if not force:
        # A repeat restore yields a SECOND live workspace holding the same
        # lineage and serial over the same real infrastructure — after which an
        # apply in either makes the other's next plan read as wholesale drift.
        # Nothing else prevents it: the marker is not consumed and `_unique_name`
        # suffixes rather than conflicting, so it used to succeed silently
        # (#1299). Refuse by default and name what already exists; `force` is
        # there for the case where the first restore was itself discarded.
        existing = await dws.read_marker(storage, workspace_id)
        prior = dws.prior_restores(existing) if existing else []
        if prior:
            raise HTTPException(
                status_code=409,
                detail=(
                    "This deletion has already been restored, into workspace(s) "
                    + ", ".join(f"ws-{i}" for i in prior)
                    + ". Restoring again would produce a second workspace holding "
                    "the same state lineage over the same infrastructure. Delete "
                    'the earlier restore first, or pass {"force": true} if it is '
                    "no longer live."
                ),
            )

    try:
        ws, report = await dws.restore_workspace(
            db,
            storage,
            workspace_id,
            restored_by=user.email,
            name=name,
            max_versions=settings.artifact_retention.state_versions_keep,
        )
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e

    if report["state_versions_restored"] == 0:
        # Nothing recoverable is a failed restore, not an empty success: the
        # state has already been reaped, or was never there. Rolling back
        # avoids leaving a bare workspace the operator then has to clean up
        # while believing their state came back.
        await db.rollback()
        raise HTTPException(
            status_code=409,
            detail=(
                "No state could be recovered for this workspace — it may already "
                "have passed its retention window and been reaped."
            ),
        )

    await db.commit()
    # After the commit, so the marker never claims a restore that rolled back.
    await dws.record_restore(
        storage, workspace_id, new_workspace_id=str(ws.id), restored_by=user.email
    )
    logger.info(
        "Restored deleted workspace",
        source_workspace_id=workspace_id,
        new_workspace_id=str(ws.id),
        new_workspace_name=ws.name,
        state_versions=report["state_versions_restored"],
        restored_by=user.email,
    )

    return {
        "data": {
            "id": f"ws-{ws.id}",
            "type": "workspaces",
            "attributes": {
                "name": ws.name,
                "restored-from": workspace_id,
                "state-versions-restored": report["state_versions_restored"],
                "state-versions-skipped": report["state_versions_skipped"],
                "suppressed": report["suppressed"],
                "dropped-references": report["dropped_references"],
            },
        }
    }
